packages/completer-extension/src/index.ts

```diff
@@ -40,23 +55,140 @@ const defaultProvider: JupyterFrontEndPlugin<void> = {
   }
 };
 
+const inlineHistoryProvider: JupyterFrontEndPlugin<void> = {
+  id: '@jupyterlab/completer-extension:inline-history',
+  description:
+    'Adds inline completion provider suggesting code from execution history.',
+  requires: [ICompletionProviderManager],
+  optional: [ITranslator],
+  autoStart: true,
+  activate: (
+    app: JupyterFrontEnd,
+    completionManager: ICompletionProviderManager,
+    translator: ITranslator | null
+  ): void => {
+    completionManager.registerInlineProvider(
+      new HistoryInlineCompletionProvider({
+        translator: translator ?? nullTranslator
+      })
+    );
+  }
+};
+
+const inlineCompleter: JupyterFrontEndPlugin<IInlineCompleterFactory> = {
+  id: '@jupyterlab/completer-extension:inline-completer',
+  description: 'Provides a factory for inline completer.',
+  provides: IInlineCompleterFactory,
+  optional: [ITranslator],
+  autoStart: true,
+  activate: (
+    app: JupyterFrontEnd,
+    translator: ITranslator | null
+  ): IInlineCompleterFactory => {
+    const trans = (translator || nullTranslator).load('jupyterlab');
+    return {
+      factory: options => {
+        const inlineCompleter = new InlineCompleter({
+          ...options,
+          trans: trans
+        });
+        inlineCompleter.toolbar.addItem(
+          'previous-inline-completion',
+          new CommandToolbarButton({
+            commands: app.commands,
+            icon: caretLeftIcon,
+            id: CommandIDs.previousInline,
+            label: trans.__('Previous')
+          })
+        );
+        inlineCompleter.toolbar.addItem(
+          'next-inline-completion',
+          new CommandToolbarButton({
+            commands: app.commands,
+            icon: caretRightIcon,
+            id: CommandIDs.nextInline,
+            label: trans.__('Next')
+          })
+        );
+        inlineCompleter.toolbar.addItem(
+          'accept-inline-completion',
+          new CommandToolbarButton({
+            commands: app.commands,
+            icon: checkIcon,
+            id: CommandIDs.acceptInline,
+            label: trans.__('Accept')
+          })
+        );
+        return inlineCompleter;
+      }
+    };
+  }
+};
+
+const inlineCompleterCommands: JupyterFrontEndPlugin<void> = {
```

Comment:
What is the reason to split this out of the previous one?

I see a risk of inconsistency if the `@jupyterlab/completer-extension:inline-completer-commands` is disabled or replace by another set of command ids.

Reply:
I am not happy with this either.

The reason is dependency chain of `inline-completer-commands` → `ICompletionProviderManager` → `IInlineCompleterFactory`. If we merge the factory with commands we have a cycle.

On the positive side, if downstream wants to override the hoverbox widget, they can just replace `IInlineCompleterFactory` and the commands on the toolbar would still be populated without them needing to cop-ypaste it so maybe it is a good thing?

Reply:
The dependency chain is now removed in 00a7f2b76fde148d3f045643632a61ddb82cbf89, the commands are now in `inline-completer` and factory is now in `inline-completer-factory`. This means that the commands are now guaranteed to be present if the inline completer is enabled, while still preserving the option to change the factory.